src/librustdoc/test.rs

```diff
@@ -348,7 +348,19 @@ pub fn make_test(s: &str,
             }
         }
     }
-    if dont_insert_main || s.contains("fn main") {
+
+    let already_has_main = s.lines()
+        .map(|line| {
+            let comment = line.find("//");
+            if let Some(comment_begins) = comment {
+                &line[0..comment_begins]
+            } else {
+                line
+            }
+        })
+        .find(|code| code.contains("fn main")).is_some();
```

Comment:
Could you make this line `.any(|code| code.contains("fn main"))` instead? That's equivalent and removes one level of method chaining.

Reply:
This is indeed a great improvement. I just have a concern on this (or the previous check):

Imagine you wrote "fn    main" (for some reasons...). Or if you write "fn main" in a string (for some other reasons...).

Reply:
True, this isn't a "complete" fix, but it's still a start. A "proper" fix would involve throwing libsyntax or the like at the doctest and scanning that for a function called "main". But for now, this is much better than we have at the moment.

Reply:
I am thinking about a partial fix involving a stupid parser creating basic tokens and all. But still bigger than this.

Reply:
I would prefer that rustdoc instead gain something like allowing `lib` where we have `rust` (i.e., in the language part after "```") today. The more work we do for the "common case" the worse rustdoc's performance on codebases with large amounts of doc tests gets.

Reply:
Agreed. There's no ideal solution in here.